pkg/identityserver/gateway_registry.go

```diff
@@ -316,11 +316,19 @@ func (is *IdentityServer) listGateways(ctx context.Context, req *ttnpb.ListGatew
 	return gtws, nil
 }
 
+func checkUpdateGatewayRights(ctx context.Context, req *ttnpb.UpdateGatewayRequest) error {
+	if len(req.FieldMask.Paths) == 1 && req.FieldMask.Paths[0] == "location" {
+		if rights.RequireGateway(ctx, req.GatewayIdentifiers, ttnpb.RIGHT_GATEWAY_LINK) == nil {
+			return nil
+		}
+	}
+	return rights.RequireGateway(ctx, req.GatewayIdentifiers, ttnpb.RIGHT_GATEWAY_SETTINGS_BASIC)
+}
+
 func (is *IdentityServer) updateGateway(ctx context.Context, req *ttnpb.UpdateGatewayRequest) (gtw *ttnpb.Gateway, err error) {
-	if err = rights.RequireGateway(ctx, req.GatewayIdentifiers, ttnpb.RIGHT_GATEWAY_SETTINGS_BASIC); err != nil {
+	if err = checkUpdateGatewayRights(ctx, req); err != nil {
 		return nil, err
 	}
-
```

Comment:
This is a bit narrow and doesn't need it's own function.

How about:

```go
func (is *IdentityServer) updateGateway(ctx context.Context, req *ttnpb.UpdateGatewayRequest) (gtw *ttnpb.Gateway, err error) {
	if err := rights.RequireGateway(ctx, req.GatewayIdentifiers, ttnpb.RIGHT_GATEWAY_SETTINGS_BASIC); err != nil {
		// Allow setting only the location field with the RIGHT_GATEWAY_LINK right.
		isLink := rights.RequireGateway(ctx, req.GatewayIdentifiers, ttnpb.RIGHT_GATEWAY_LINK) == nil
		if topLevel := ttnpb.TopLevelFields(req.FieldMask.Paths); len(topLevel) != 1 || topLevel[0] != "location" || !isLink {
			return nil, err
		}
	}
```

Reply:
The point is to preserve the error message being about the gateway settings rights, not the gateway link rights.

Seems a tad more sane to me, but what you are suggesting is also fine if we do not really care about that.

Reply:
> The point is to preserve the error message being about the gateway settings rights, not the gateway link rights.

Did you see the code above?

Reply:
Ah, missed the `isLink`.